src/main.py

```diff
@@ -93,44 +97,101 @@ def main() -> None:
-            learning_time = time.time() - learning_start
+        teacher = Teacher()
+        for _ in range(CFG.num_online_learning_cycles):
             results = _run_testing(env, approach)
-            _save_test_results(results, learning_time=learning_time)
+            _save_test_results(results, num_transitions=total_num_transitions,
+                               learning_time=(time.time()-learning_start))
+            requests = approach.get_interaction_requests()
+            if not requests:
+                break
+            responses, new_num_transitions = _generate_interaction_responses(
+                env.simulate, teacher, train_tasks, requests)
+            total_num_transitions += new_num_transitions
+            approach.learn_from_interaction_responses(responses)
     else:
         results = _run_testing(env, approach)
-        _save_test_results(results, learning_time=0.0)
+        _save_test_results(results, num_transitions=0, learning_time=0.0)
     script_time = time.time() - script_start
     print(f"\n\nMain script terminated in {script_time:.5f} seconds")
 
 
+def _generate_or_load_offline_dataset(env: BaseEnv, train_tasks: List[Task]
+                                      ) -> Tuple[Dataset, int]:
+    """Create offline dataset from training tasks. Returns both the dataset
+    and its size (the number of transitions in all the data).
+    """
+    dataset_filename = (
+        f"{CFG.env}__{CFG.offline_data_method}__{CFG.seed}.data")
+    dataset_filepath = os.path.join(CFG.data_dir, dataset_filename)
+    if CFG.load_data:
+        assert os.path.exists(dataset_filepath)
+        with open(dataset_filepath, "rb") as f:
+            dataset = pkl.load(f)
+        print("\n\nLOADED DATASET")
+    else:
+        dataset = create_dataset(env, train_tasks)
+        print("\n\nCREATED DATASET")
+        os.makedirs(CFG.data_dir, exist_ok=True)
+        with open(dataset_filepath, "wb") as f:
+            pkl.dump(dataset, f)
+    num_transitions = 0
+    for traj in dataset:
+        num_transitions += len(traj.actions)
+    return dataset, num_transitions
+
+
+def _generate_interaction_responses(
+        simulator: Callable[[State, Action], State], teacher: Teacher,
+        train_tasks: List[Task], requests: Sequence[InteractionRequest]
+) -> Tuple[List[InteractionResponse], int]:
+    """Given a sequence of InteractionRequest objects, return a list of
+    InteractionResponse objects and the total number of transitions.
+    """
+    interaction_responses = []
+    num_transitions = 0
+    for request in requests:
+        # First, roll out the acting policy.
+        task = train_tasks[request.train_task_idx]
+        traj = utils.run_policy_until(
+            request.act_policy, simulator, task.init,
+            request.termination_function,
+            max_num_steps=CFG.max_num_steps_interaction_request)
+        # Now, go through the trajectory and handle queries
```

Comment:
can you add a note "apologizing" for the two passes through the states (because of the disconnect with reality)?

Reply:
it's not bc of any disconnect from reality, it's just bc i wanted to use `utils.run_policy_until` without modifying it lol. you could always make a new function that steps the policy through and does the query handling at the same time.